### ISSUE TITLE
Fitting LGBMClassifier on a Dataset

### ISSUE THREAD

**Issue by sh0gg0th** [microsoft/LightGBM#5074]
Using the sklearn API, I'm currently training LGBMClassifier like this:
```
train_ds = lgbm.Dataset(X_train, label=y_train)
model = lgbm.LGBMClassifier(...)
model.fit(train_ds.data, train_ds.label, ...)
```
Do I incur any memory or performance penalties with `train_ds.data`, `train_ds.label`? My data is very large - tens of gigabytes. I know I can run `lightgbm.train` directly on a Dataset, but it doesn't have the `early_stopping` callback that I need.

In general, what are the advantages of using a Dataset over a numpy array?

**Comment by jameslamb** [comment]
Thanks for using LightGBM! 

> what are the advantages of using a Dataset over a numpy array?

Please see [this stack overflow answer](https://stackoverflow.com/a/66004139/3986677) to the same question.

In short, LightGBM does some pre-processing of the raw data prior to training, and it's that preprocessed representation that is used as the input to boosting.

The scikit estimators like `LGBMClassifier` and `LGBMRegressor` are just intended to provide a way to train a LightGBM model that is compatible with the main other libraries that understand estimators following the `scikit-learn` API. When you provide data in `numpy`, `scipy`, or `pandas` format to `LGBMClassifier.fit()`, internally the `lightgbm` Python package is creating a `Dataset` from that raw data.

Don't build a `Dataset` and then pass its `.data` and `.label` to `LGBMClassifier.fit()`.

Use one of these options:

* create a `Dataset` using `lgb.Dataset()`, set `free_raw_data=True` to remove the raw data after construction, then use `lgb.train()` to train a model
* use `LGBMClassifier()` on data in `numpy` format, and don't create a `Dataset` at all.

> My data is very large - tens of gigabytes

If you don't have sufficient memory to hold your raw data and the `lgb.Dataset` object in memory at the same time, please consider using the `Sequence` API added in #4089: https://github.com/microsoft/LightGBM/blob/f834cab49a3e394969d087bb209d500b70f428dd/examples/python-guide/dataset_from_multi_hdf5.py

That API allows the construction of `lgb.Dataset` from chunks of the full dataset, which are read into memory and processed one at a time.

Alternatively, if you're able to provision a Dask cluster, you could try distributed learning with `lightgbm.dask` (https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#dask).

**Comment by sh0gg0th** [comment]
Thanks for a very quick reply, @jameslamb ! I do appear to have enough memory, but I'm running out when I'm wrapping my training in an Optuna `objective`. My understanding is that out of your suggested options, I should use
> create a Dataset using lgb.Dataset(), set free_raw_data=True
to avoid preprocessing of the same raw data at each Optuna trial.

However, I don't see any references to early stopping in [the training  API](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html), should I be looking elsewhere?

**Comment by jameslamb** [comment]
> However, I don't see any references to early_stopping_rounds in the [training API](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html), should I be looking elsewhere?

The current supported way to perform early stopping using `lgb.train()` is to use the `lgb.early_stopping()` callback, like this:

https://github.com/microsoft/LightGBM/blob/f834cab49a3e394969d087bb209d500b70f428dd/examples/python-guide/simple_example.py#L39-L43

I recommend using that approach, as the old option of providing `early_stopping_rounds` as a keyword argument to `lgb.train()` (https://lightgbm.readthedocs.io/en/v3.3.2/pythonapi/lightgbm.train.html#lightgbm.train) was marked deprecated as of release v3.3.1 and removed in #4908.

**Comment by sh0gg0th** [comment]
Understood, thanks a lot!

**Comment by jameslamb** [comment]
No prob, thanks for using LightGBM!
